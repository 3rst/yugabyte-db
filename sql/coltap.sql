\unset ECHO
\i test_setup.sql

-- $Id$

SELECT plan(135);
--SELECT * from no_plan();

-- This will be rolled back. :-)
SET client_min_messages = warning;
CREATE TABLE public.sometab(
    id    INT NOT NULL PRIMARY KEY,
    name  TEXT DEFAULT '',
    numb  NUMERIC(10, 2) DEFAULT NULL,
    myint NUMERIC(8) DEFAULT 24,
    plain INTEGER
);
RESET client_min_messages;

/****************************************************************************/
-- Test col_not_null().
SELECT * FROM check_test(
    col_not_null( 'pg_catalog', 'pg_type', 'typname', 'typname not null' ),
    true,
    'col_not_null( sch, tab, col, desc )',
    'typname not null',
    ''
);

SELECT * FROM check_test(
    col_not_null( 'sometab', 'id', 'blah blah blah' ),
    true,
    'col_not_null( tab, col, desc )',
    'blah blah blah',
    ''
);

SELECT * FROM check_test(
    col_not_null( 'sometab', 'id' ),
    true,
    'col_not_null( table, column )',
    'Column sometab.id should be NOT NULL',
    ''
);

-- Make sure failure is correct.
SELECT * FROM check_test(
    col_not_null( 'sometab', 'name' ),
    false,
    'col_not_null( table, column ) fail',
    'Column sometab.name should be NOT NULL',
    ''
);

/****************************************************************************/
-- Test col_is_null().
SELECT * FROM check_test(
    col_is_null( 'public', 'sometab', 'name', 'name is null' ),
    true,
    'col_is_null( sch, tab, col, desc )',
    'name is null',
    ''
);

SELECT * FROM check_test(
    col_is_null( 'sometab', 'name', 'my desc' ),
    true,
    'col_is_null( tab, col, desc )',
    'my desc',
    ''
);

SELECT * FROM check_test(
    col_is_null( 'sometab', 'name' ),
    true,
    'col_is_null( tab, col )',
    'Column sometab.name should allow NULL',
    ''
);
-- Make sure failure is correct.
SELECT * FROM check_test(
    col_is_null( 'sometab', 'id' ),
    false,
    'col_is_null( tab, col ) fail',
    'Column sometab.id should allow NULL',
    ''
);

/****************************************************************************/
-- Test col_type_is().
SELECT * FROM check_test(
    col_type_is( 'public', 'sometab', 'name', 'text', 'name is text' ),
    true,
    'col_type_is( sch, tab, col, type, desc )',
    'name is text',
    ''
);

SELECT * FROM check_test(
    col_type_is( 'sometab', 'name', 'text', 'yadda yadda yadda' ),
    true,
    'col_type_is( tab, col, type, desc )',
    'yadda yadda yadda',
    ''
);

SELECT * FROM check_test(
    col_type_is( 'sometab', 'name', 'text' ),
    true,
    'col_type_is( tab, col, type )',
    'Column sometab.name should be type text',
    ''
);

SELECT * FROM check_test(
    col_type_is( 'sometab', 'name', 'TEXT' ),
    true,
    'col_type_is( tab, col, type ) insensitive',
    'Column sometab.name should be type TEXT',
    ''
);

-- Make sure failure is correct.
SELECT * FROM check_test(
    col_type_is( 'sometab', 'name', 'int4' ),
    false,
    'col_type_is( tab, col, type ) fail',
    'Column sometab.name should be type int4',
    '        have: text
        want: int4'
);

/****************************************************************************/
-- Try col_type_is() with precision.
SELECT * FROM check_test(
    col_type_is( 'public', 'sometab', 'numb', 'numeric(10,2)', 'lol' ),
    true,
    'col_type_is with precision',
    'lol',
    ''
);

-- Check its diagnostics.
SELECT * FROM check_test(
    col_type_is( 'sometab', 'myint', 'numeric(7)', 'should be numeric(7)' ),
    false,
    'col_type_is precision fail',
    'should be numeric(7)',
    '        have: numeric(8,0)
        want: numeric(7)'
);

/****************************************************************************/
-- Test col_has_default().
SELECT * FROM check_test(
    col_has_default( 'public', 'sometab', 'name', 'desc' ),
    true,
    'col_has_default( sch, tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_has_default( 'sometab', 'name', 'desc' ),
    true,
    'col_has_default( tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_has_default( 'sometab', 'name' ),
    true,
    'col_has_default( tab, col )',
    'Column sometab.name should have a default',
    ''
);

-- Check with a column with no default.
SELECT * FROM check_test(
    col_has_default( 'public', 'sometab', 'plain', 'desc' ),
    false,
    'col_has_default( sch, tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_has_default( 'sometab', 'plain', 'desc' ),
    false,
    'col_has_default( tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_has_default( 'sometab', 'plain' ),
    false,
    'col_has_default( tab, col )',
    'Column sometab.plain should have a default',
    ''
);

-- Check with a nonexistent column.
SELECT * FROM check_test(
    col_has_default( 'public', 'sometab', '__asdfasdfs__', 'desc' ),
    false,
    'col_has_default( sch, tab, col, desc )',
    'desc',
    '    Column public.sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_has_default( 'sometab', '__asdfasdfs__', 'desc' ),
    false,
    'col_has_default( tab, col, desc )',
    'desc',
    '    Column sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_has_default( 'sometab', '__asdfasdfs__' ),
    false,
    'col_has_default( tab, col )',
    'Column sometab.__asdfasdfs__ should have a default',
    '    Column sometab.__asdfasdfs__ does not exist'
);

/****************************************************************************/
-- Test col_hasnt_default().
SELECT * FROM check_test(
    col_hasnt_default( 'public', 'sometab', 'name', 'desc' ),
    false,
    'col_hasnt_default( sch, tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', 'name', 'desc' ),
    false,
    'col_hasnt_default( tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', 'name' ),
    false,
    'col_hasnt_default( tab, col )',
    'Column sometab.name should not have a default',
    ''
);

-- Check with a column with no default.
SELECT * FROM check_test(
    col_hasnt_default( 'public', 'sometab', 'plain', 'desc' ),
    true,
    'col_hasnt_default( sch, tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', 'plain', 'desc' ),
    true,
    'col_hasnt_default( tab, col, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', 'plain' ),
    true,
    'col_hasnt_default( tab, col )',
    'Column sometab.plain should not have a default',
    ''
);

-- Check with a nonexistent column.
SELECT * FROM check_test(
    col_hasnt_default( 'public', 'sometab', '__asdfasdfs__', 'desc' ),
    false,
    'col_hasnt_default( sch, tab, col, desc )',
    'desc',
    '    Column public.sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', '__asdfasdfs__', 'desc' ),
    false,
    'col_hasnt_default( tab, col, desc )',
    'desc',
    '    Column sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_hasnt_default( 'sometab', '__asdfasdfs__' ),
    false,
    'col_hasnt_default( tab, col )',
    'Column sometab.__asdfasdfs__ should not have a default',
    '    Column sometab.__asdfasdfs__ does not exist'
);

/****************************************************************************/
-- Test col_default_is().

SELECT * FROM check_test(
    col_default_is( 'public', 'sometab', 'name', ''::text, 'name should default to empty string' ),
    true,
    'col_default_is( sch, tab, col, def, desc )',
    'name should default to empty string',
    ''
);

SELECT * FROM check_test(
    col_default_is( 'public', 'sometab', 'name', 'foo'::text, 'name should default to ''foo''' ),
    false,
    'col_default_is() fail',
    'name should default to ''foo''',
    '        have: 
        want: foo'
);

SELECT * FROM check_test(
    col_default_is( 'sometab', 'name', ''::text, 'name should default to empty string' ),
    true,
    'col_default_is( tab, col, def, desc )',
    'name should default to empty string',
    ''
);

SELECT * FROM check_test(
    col_default_is( 'sometab', 'name', ''::text ),
    true,
    'col_default_is( tab, col, def )',
    'Column sometab.name should default to ''''',
    ''
);

-- Make sure it works with a non-text column.
SELECT * FROM check_test(
    col_default_is( 'sometab', 'myint', 24 ),
    true,
    'col_default_is( tab, col, int )',
    'Column sometab.myint should default to ''24''',
    ''
);

-- Make sure it works with a NULL default.
SELECT * FROM check_test(
    col_default_is( 'sometab', 'numb', NULL::numeric, 'desc' ),
    true,
    'col_default_is( tab, col, NULL, desc )',
    'desc',
    ''
);
SELECT * FROM check_test(
    col_default_is( 'sometab', 'numb', NULL::numeric ),
    true,
    'col_default_is( tab, col, NULL )',
    'Column sometab.numb should default to NULL',
    ''
);

-- Make sure that it fails when there is no default.
SELECT * FROM check_test(
    col_default_is( 'sometab', 'plain', 1::integer, 'desc' ),
    false,
    'col_default_is( tab, col, bogus, desc )',
    'desc',
    '    Column sometab.plain has no default'
);

SELECT * FROM check_test(
    col_default_is( 'sometab', 'plain', 1::integer ),
    false,
    'col_default_is( tab, col, bogus )',
    'Column sometab.plain should default to ''1''',
    '    Column sometab.plain has no default'
);

-- Check with a nonexistent column.
SELECT * FROM check_test(
    col_default_is( 'public', 'sometab', '__asdfasdfs__', NULL::text, 'desc' ),
    false,
    'col_default_is( sch, tab, col, def, desc )',
    'desc',
    '    Column public.sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_default_is( 'sometab', '__asdfasdfs__', NULL::text, 'desc' ),
    false,
    'col_default_is( tab, col, def, desc )',
    'desc',
    '    Column sometab.__asdfasdfs__ does not exist'
);
SELECT * FROM check_test(
    col_default_is( 'sometab', '__asdfasdfs__', NULL::text ),
    false,
    'col_default_is( tab, col, def )',
    'Column sometab.__asdfasdfs__ should default to NULL',
    '    Column sometab.__asdfasdfs__ does not exist'
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
