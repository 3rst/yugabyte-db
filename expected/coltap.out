\unset ECHO
1..153
ok 1 - col_not_null( sch, tab, col, desc ) should pass
ok 2 - col_not_null( sch, tab, col, desc ) should have the proper description
ok 3 - col_not_null( sch, tab, col, desc ) should have the proper diagnostics
ok 4 - col_not_null( tab, col, desc ) should pass
ok 5 - col_not_null( tab, col, desc ) should have the proper description
ok 6 - col_not_null( tab, col, desc ) should have the proper diagnostics
ok 7 - col_not_null( table, column ) should pass
ok 8 - col_not_null( table, column ) should have the proper description
ok 9 - col_not_null( table, column ) should have the proper diagnostics
ok 10 - col_not_null( table, column ) fail should fail
ok 11 - col_not_null( table, column ) fail should have the proper description
ok 12 - col_not_null( table, column ) fail should have the proper diagnostics
ok 13 - col_not_null( sch, tab, noncol, desc ) should fail
ok 14 - col_not_null( sch, tab, noncol, desc ) should have the proper description
ok 15 - col_not_null( sch, tab, noncol, desc ) should have the proper diagnostics
ok 16 - col_not_null( table, noncolumn ) fail should fail
ok 17 - col_not_null( table, noncolumn ) fail should have the proper description
ok 18 - col_not_null( table, noncolumn ) fail should have the proper diagnostics
ok 19 - col_is_null( sch, tab, col, desc ) should pass
ok 20 - col_is_null( sch, tab, col, desc ) should have the proper description
ok 21 - col_is_null( sch, tab, col, desc ) should have the proper diagnostics
ok 22 - col_is_null( tab, col, desc ) should pass
ok 23 - col_is_null( tab, col, desc ) should have the proper description
ok 24 - col_is_null( tab, col, desc ) should have the proper diagnostics
ok 25 - col_is_null( tab, col ) should pass
ok 26 - col_is_null( tab, col ) should have the proper description
ok 27 - col_is_null( tab, col ) should have the proper diagnostics
ok 28 - col_is_null( tab, col ) fail should fail
ok 29 - col_is_null( tab, col ) fail should have the proper description
ok 30 - col_is_null( tab, col ) fail should have the proper diagnostics
ok 31 - col_is_null( sch, tab, noncol, desc ) should fail
ok 32 - col_is_null( sch, tab, noncol, desc ) should have the proper description
ok 33 - col_is_null( sch, tab, noncol, desc ) should have the proper diagnostics
ok 34 - col_is_null( table, noncolumn ) fail should fail
ok 35 - col_is_null( table, noncolumn ) fail should have the proper description
ok 36 - col_is_null( table, noncolumn ) fail should have the proper diagnostics
ok 37 - col_type_is( sch, tab, col, type, desc ) should pass
ok 38 - col_type_is( sch, tab, col, type, desc ) should have the proper description
ok 39 - col_type_is( sch, tab, col, type, desc ) should have the proper diagnostics
ok 40 - col_type_is( tab, col, type, desc ) should pass
ok 41 - col_type_is( tab, col, type, desc ) should have the proper description
ok 42 - col_type_is( tab, col, type, desc ) should have the proper diagnostics
ok 43 - col_type_is( tab, col, type ) should pass
ok 44 - col_type_is( tab, col, type ) should have the proper description
ok 45 - col_type_is( tab, col, type ) should have the proper diagnostics
ok 46 - col_type_is( tab, col, type ) insensitive should pass
ok 47 - col_type_is( tab, col, type ) insensitive should have the proper description
ok 48 - col_type_is( tab, col, type ) insensitive should have the proper diagnostics
ok 49 - col_type_is( tab, col, type ) fail should fail
ok 50 - col_type_is( tab, col, type ) fail should have the proper description
ok 51 - col_type_is( tab, col, type ) fail should have the proper diagnostics
ok 52 - col_type_is( tab, noncol, type ) fail should fail
ok 53 - col_type_is( tab, noncol, type ) fail should have the proper description
ok 54 - col_type_is( tab, noncol, type ) fail should have the proper diagnostics
ok 55 - col_type_is( sch, tab, noncol, type, desc ) fail should fail
ok 56 - col_type_is( sch, tab, noncol, type, desc ) fail should have the proper description
ok 57 - col_type_is( sch, tab, noncol, type, desc ) fail should have the proper diagnostics
ok 58 - col_type_is with precision should pass
ok 59 - col_type_is with precision should have the proper description
ok 60 - col_type_is with precision should have the proper diagnostics
ok 61 - col_type_is precision fail should fail
ok 62 - col_type_is precision fail should have the proper description
ok 63 - col_type_is precision fail should have the proper diagnostics
ok 64 - col_has_default( sch, tab, col, desc ) should pass
ok 65 - col_has_default( sch, tab, col, desc ) should have the proper description
ok 66 - col_has_default( sch, tab, col, desc ) should have the proper diagnostics
ok 67 - col_has_default( tab, col, desc ) should pass
ok 68 - col_has_default( tab, col, desc ) should have the proper description
ok 69 - col_has_default( tab, col, desc ) should have the proper diagnostics
ok 70 - col_has_default( tab, col ) should pass
ok 71 - col_has_default( tab, col ) should have the proper description
ok 72 - col_has_default( tab, col ) should have the proper diagnostics
ok 73 - col_has_default( sch, tab, col, desc ) should fail
ok 74 - col_has_default( sch, tab, col, desc ) should have the proper description
ok 75 - col_has_default( sch, tab, col, desc ) should have the proper diagnostics
ok 76 - col_has_default( tab, col, desc ) should fail
ok 77 - col_has_default( tab, col, desc ) should have the proper description
ok 78 - col_has_default( tab, col, desc ) should have the proper diagnostics
ok 79 - col_has_default( tab, col ) should fail
ok 80 - col_has_default( tab, col ) should have the proper description
ok 81 - col_has_default( tab, col ) should have the proper diagnostics
ok 82 - col_has_default( sch, tab, col, desc ) should fail
ok 83 - col_has_default( sch, tab, col, desc ) should have the proper description
ok 84 - col_has_default( sch, tab, col, desc ) should have the proper diagnostics
ok 85 - col_has_default( tab, col, desc ) should fail
ok 86 - col_has_default( tab, col, desc ) should have the proper description
ok 87 - col_has_default( tab, col, desc ) should have the proper diagnostics
ok 88 - col_has_default( tab, col ) should fail
ok 89 - col_has_default( tab, col ) should have the proper description
ok 90 - col_has_default( tab, col ) should have the proper diagnostics
ok 91 - col_hasnt_default( sch, tab, col, desc ) should fail
ok 92 - col_hasnt_default( sch, tab, col, desc ) should have the proper description
ok 93 - col_hasnt_default( sch, tab, col, desc ) should have the proper diagnostics
ok 94 - col_hasnt_default( tab, col, desc ) should fail
ok 95 - col_hasnt_default( tab, col, desc ) should have the proper description
ok 96 - col_hasnt_default( tab, col, desc ) should have the proper diagnostics
ok 97 - col_hasnt_default( tab, col ) should fail
ok 98 - col_hasnt_default( tab, col ) should have the proper description
ok 99 - col_hasnt_default( tab, col ) should have the proper diagnostics
ok 100 - col_hasnt_default( sch, tab, col, desc ) should pass
ok 101 - col_hasnt_default( sch, tab, col, desc ) should have the proper description
ok 102 - col_hasnt_default( sch, tab, col, desc ) should have the proper diagnostics
ok 103 - col_hasnt_default( tab, col, desc ) should pass
ok 104 - col_hasnt_default( tab, col, desc ) should have the proper description
ok 105 - col_hasnt_default( tab, col, desc ) should have the proper diagnostics
ok 106 - col_hasnt_default( tab, col ) should pass
ok 107 - col_hasnt_default( tab, col ) should have the proper description
ok 108 - col_hasnt_default( tab, col ) should have the proper diagnostics
ok 109 - col_hasnt_default( sch, tab, col, desc ) should fail
ok 110 - col_hasnt_default( sch, tab, col, desc ) should have the proper description
ok 111 - col_hasnt_default( sch, tab, col, desc ) should have the proper diagnostics
ok 112 - col_hasnt_default( tab, col, desc ) should fail
ok 113 - col_hasnt_default( tab, col, desc ) should have the proper description
ok 114 - col_hasnt_default( tab, col, desc ) should have the proper diagnostics
ok 115 - col_hasnt_default( tab, col ) should fail
ok 116 - col_hasnt_default( tab, col ) should have the proper description
ok 117 - col_hasnt_default( tab, col ) should have the proper diagnostics
ok 118 - col_default_is( sch, tab, col, def, desc ) should pass
ok 119 - col_default_is( sch, tab, col, def, desc ) should have the proper description
ok 120 - col_default_is( sch, tab, col, def, desc ) should have the proper diagnostics
ok 121 - col_default_is() fail should fail
ok 122 - col_default_is() fail should have the proper description
ok 123 - col_default_is() fail should have the proper diagnostics
ok 124 - col_default_is( tab, col, def, desc ) should pass
ok 125 - col_default_is( tab, col, def, desc ) should have the proper description
ok 126 - col_default_is( tab, col, def, desc ) should have the proper diagnostics
ok 127 - col_default_is( tab, col, def ) should pass
ok 128 - col_default_is( tab, col, def ) should have the proper description
ok 129 - col_default_is( tab, col, def ) should have the proper diagnostics
ok 130 - col_default_is( tab, col, int ) should pass
ok 131 - col_default_is( tab, col, int ) should have the proper description
ok 132 - col_default_is( tab, col, int ) should have the proper diagnostics
ok 133 - col_default_is( tab, col, NULL, desc ) should pass
ok 134 - col_default_is( tab, col, NULL, desc ) should have the proper description
ok 135 - col_default_is( tab, col, NULL, desc ) should have the proper diagnostics
ok 136 - col_default_is( tab, col, NULL ) should pass
ok 137 - col_default_is( tab, col, NULL ) should have the proper description
ok 138 - col_default_is( tab, col, NULL ) should have the proper diagnostics
ok 139 - col_default_is( tab, col, bogus, desc ) should fail
ok 140 - col_default_is( tab, col, bogus, desc ) should have the proper description
ok 141 - col_default_is( tab, col, bogus, desc ) should have the proper diagnostics
ok 142 - col_default_is( tab, col, bogus ) should fail
ok 143 - col_default_is( tab, col, bogus ) should have the proper description
ok 144 - col_default_is( tab, col, bogus ) should have the proper diagnostics
ok 145 - col_default_is( sch, tab, col, def, desc ) should fail
ok 146 - col_default_is( sch, tab, col, def, desc ) should have the proper description
ok 147 - col_default_is( sch, tab, col, def, desc ) should have the proper diagnostics
ok 148 - col_default_is( tab, col, def, desc ) should fail
ok 149 - col_default_is( tab, col, def, desc ) should have the proper description
ok 150 - col_default_is( tab, col, def, desc ) should have the proper diagnostics
ok 151 - col_default_is( tab, col, def ) should fail
ok 152 - col_default_is( tab, col, def ) should have the proper description
ok 153 - col_default_is( tab, col, def ) should have the proper diagnostics
